pokedex/simulator/pokedex/src/simulator.rs

```diff
@@ -359,3 +335,314 @@ impl Drop for SimulatorHandle {
+    }
+
+    fn get_value(&self) -> [u8; 4] {
+        let v = self.1.load(Ordering::Acquire);
+        v.to_le_bytes()
+    }
+}
+
+impl BusInfo {
+    pub fn from_device_config<P: AsRef<Path> + Debug>(p: P) -> miette::Result<Self> {
+        let content = std::fs::read_to_string(&p)
+            .into_diagnostic()
+            .wrap_err_with(|| format!("fail read path: {:?}", &p))?;
+
+        let filepath = p.as_ref().to_string_lossy();
+        let configuration: Vec<AddressSpaceDescNode> = knuffel::parse(&filepath, &content)?;
+
+        let mut segments = Vec::new();
+        let mut ic = None;
+        for node in configuration {
+            match node {
+                AddressSpaceDescNode::SRAM(sram) => {
+                    let naive_memory = NaiveMemory::new(sram.length as usize);
+                    let boxed: Box<dyn Addressable> = Box::new(naive_memory);
+                    segments.push(((sram.base..sram.base + sram.length), boxed))
+                }
+                AddressSpaceDescNode::MMIO(mmio_config) => {
+                    let (mmio_decoder, controllers) =
+                        MMIOAddrDecoder::try_build_from(mmio_config.mmap)?;
+                    ic = Some(controllers);
+                    let boxed: Box<dyn Addressable> = Box::new(mmio_decoder);
+                    segments.push((
+                        (mmio_config.base..mmio_config.base + mmio_config.length),
+                        boxed,
+                    ))
+                }
+            }
+        }
+
+        let mut overlapped_segment = None;
+        let mut unchecked_index: Vec<Range<u32>> =
+            segments.iter().map(|(index, _)| index.clone()).collect();
+        unchecked_index.sort_by_key(|range| range.start);
+        for window in unchecked_index.windows(2) {
+            let addr1 = &window[0];
+            let addr2 = &window[1];
+            if addr2.start < addr1.end {
+                overlapped_segment = Some(&window[1]);
+            }
+        }
+
+        if let Some(range) = overlapped_segment {
+            miette::bail!(
+                "Address space with offset={:#010x} length={:#010x} overlapped previous address space",
+                range.start,
+                range.end - range.start
+            )
+        }
+
+        Ok(Self {
+            interrupt_controllers: ic.unwrap(),
+            bus_bridge: BusBridge {
+                address_space: segments,
+            },
+        })
+    }
+}
+
+#[test]
+fn test_address_space() -> miette::Result<()> {
+    let _ = BusInfo::from_device_config("./assets/devices.kdl")?;
+    Ok(())
+}
+
+#[derive(Debug)]
+pub enum BusError {
+    OutOfMemory {
+        id: &'static str,
+        addr: u32,
+        len: u32,
+    },
+    Illegal,
+}
+
+pub trait Addressable: std::fmt::Debug {
+    fn resp_bus_read(&self, offset: u32, dest: &mut [u8]) -> Result<(), BusError>;
+    fn resp_bus_write(&mut self, offset: u32, data: &[u8]) -> Result<(), BusError>;
+}
```

Comment:
1. may simply use `bus_read`/`bus_write`
2. MMIO read may have side effects. It should take `&mut self`
3. `Debug` super trait seems not very useful. It seems we don't want to dump the whole memory. If it's not used, we may remove it.
4. We may add `Send` trait bound somewhere, ensure it could be wrapped inside `Mutex` in the future.

Reply:
In my opinion methods naming is better to start with verb, and each read/write to a MMIO device is actually require a response from them.

Reply:
> In my opinion methods naming is better to start with verb

If so, may use `do_bus_read`

> and each read/write to a MMIO device is actually requiring a response from them.

The response is exactly the return value. The function call perform both request and response, only name it 'resp' is misleading.